add support for modifying operators

djerius opened the pull request:
This changeset adds support for modifying operator overloads.  For overloads which are coderefs rather than method names, it creates a named entry in the package's symbol table for the coderef and updates the overloaded operator with that name.  That is then used by the modifiers.

Unfortunately there is now a dependency on mro/MRO::Compat.  Using the "can" method on the overload entries in the symbol table messes up the overload mechanism (see overload::mycan, which since 5.10 uses mro).  I could switch the code to use overload::mycan, but I'll be honest in that I don't know what affect that'll have on pre-mro Perls.

djerius commented:
I've created another branch, [operators-no-mro](https://github.com/djerius/Class-Method-Modifiers/tree/operators-no-mro) which uses overload::mycan instead of mro/MRO::Compat

haarg commented on lib/Class/Method/Modifiers.pm:
```diff
@@ -18,8 +18,15 @@ our %EXPORT_TAGS = (
 
 BEGIN {
   *_HAS_READONLY = $] >= 5.008 ? sub(){1} : sub(){0};
+  *_MRO_MODULE   = "$]" < 5.010 ? sub() { "MRO/Compat.pm" } : sub() { "mro.pm" };
```
We'd definitely want to go with the variant of this that doesn't require MRO::Compat.

haarg commented on lib/Class/Method/Modifiers.pm:
```diff
@@ -18,8 +18,15 @@ our %EXPORT_TAGS = (
 
+# generate list of overloadable operators
+use overload ();
+my %OP = map {; $_ => 1 } map { split( ' ', $overload::ops{$_} ) } keys %overload::ops;
```
`grep $_ ne 'fallback', map split(/\s+/, $_), values %overload::ops`

djerius replied:
ouch.  that's much cleaner, exclusion of fallback noted.

haarg commented on lib/Class/Method/Modifiers.pm:
```diff
@@ -37,9 +44,15 @@ sub install_modifier {
+        my $orig = $name;
+
+        $name = _overloaded_op_method( $into, $name )
+          if $OP{$name};
```
This may be reasonable for things like `""` that will never be a normal method name, it isn't workable for those that are valid identifiers (`cmp`, `sin`, `bool`, etc).

haarg commented on lib/Class/Method/Modifiers.pm:
```diff
@@ -218,6 +231,84 @@ sub _is_in_package {
+        my $glob = _getglob "${class}::${symbol}";
+        *$glob = \$method_name;
+        no warnings 'redefine';
+        *$glob = \&overload::nil || \&overload::_nil;
```
There was an idea to change `nil` to be `_nil` at some point that never went anywhere afaik.  We may want to just forget about supporting that.  If we are supporting that though, this won't work.  `\&overload::nil` will force the code-ref to exist and will never be false.  It needs a defined check like is used above.

shadowcat-mst commented:
What's the use case for this? Given sane overloading code that does

    use overload '""' => 'stringify', ...;

you can just apply modifiers to the stringify method.

If you use the form of overload that takes direct coderefs, you are explicitly saying that you don't want to co-operate with subclassing.

It seems strange to me to be able to 'around' something that you can't subclass normally?

haarg commented:
I'm also realizing that with this implementation,
```
use overload '""' => 'as_string';
sub as_string { ... }
around '""' => sub { ... };
```
would apply a modifier to `as_string`.  I don't think that's sensible at all.

djerius commented:
@shadowcat-mst  I'm working with a decades old codebase (PDL) which doesn't follow normal practice.  There is a form of [subclassing](https://metacpan.org/pod/release/CHM/PDL-2.017/Basic/Pod/Objects.pod)  All of the operator overloads are done with direct [coderefs](https://metacpan.org/source/CHM/PDL-2.017/Basic/Core/Core.pm) rather than methods.  It's a historical oddity.  Not all of the design decisions were informed ones. There's no chance that the existing codebase will be cleaned up.

shadowcat-mst commented:
@djerius right, well, 'make a mess of CMM' isn't really a good way to work around PDL being ancient.

How about writing a module called something like overload::reify that takes *all* the coderef-based overloads and turns them into methods in a predictable fashion, and then you can just use plain CMM on top of that? That way you get your sanity, the process of extracting sanity from old overload code is available to everybody, even those not using CMM, and CMM doesn't end up with weird code in it that doesn't belong there?

(and you can totally submit doc matches mentioning said overload::reify or whatever to Moo, Moose et. al. so anybody with this problem can share the solution rather than only CMM using modules, which Moose actually isn't)

karenetheridge commented:
That's a really good point -- if the code to make overloads modifiable was in its own distribution, then Moose can make use of it too (since it has its own implementation of method overloads, separate from Class::Method::Modifiers).

karenetheridge commented:
That's a really good point -- if the code to make overloads modifiable was in its own distribution, then Moose can make use of it too (since it has its own implementation of method overloads, separate from Class::Method::Modifiers).

djerius commented:
Thanks for all of your input.  As per the suggestion of @shadowcat-mst I've put the code in a new module:[overload::reify](https://github.com/djerius/overload-reify).  I hope it addresses all of your concerns.
